src/thread/srp.rs

```diff
@@ -0,0 +1,1072 @@
+
+    /// Remove the SRP hostname and all SRP services from the SRP client.
+    ///
+    /// Arguments:
+    /// - `immediate`: If `true`, the hostname and services will be removed immediately, otherwise,
+    ///   the hostname and services will be removed gracefully by propagating the removal info to the SRP server.
+    pub fn srp_remove_all(&self, immediate: bool) -> Result<(), EspError> {
+        self.driver().srp_remove_all(immediate)
+    }
+
+    // /// Wait for the SRP state to change.
+    // ///
+    // /// This method will wait forever if `OpenThread` is not instantiated with SRP.
+    // ///
+    // /// NOTE:
+    // /// It is not advised to call this method concurrently from multiple async tasks
+    // /// because it uses a single waker registration. Thus, while the method will not panic,
+    // /// the tasks will fight with each other by each re-registering its own waker, thus keeping the CPU constantly busy.
+    // TODO
+    // pub async fn srp_wait_changed(&self) {
+    //     self.driver().srp_wait_changed().await
+    // }
+}
+
+// TODO: Make these configurable with a feature
+const SRP_SVCS: usize = 3;
+const SRP_SVC_BUF_SIZE: usize = 300;
+const SRP_HOST_BUF_SIZE: usize = 300;
+
+pub(crate) struct OtSrp {
+    conf_taken: bool,
+    conf_buf: [u8; SRP_HOST_BUF_SIZE],
+    services: [OtSrpService; SRP_SVCS],
+}
+
+impl OtSrp {
+    pub(crate) unsafe fn init(this: *mut Self) {
+        unsafe {
+            addr_of_mut!((*this).conf_taken).write(false);
+            addr_of_mut!((*this).conf_buf).write_bytes(0, 1);
+
+            for index in 0..SRP_SVCS {
+                let service = addr_of_mut!((*this).services[index]);
+                OtSrpService::init(service);
+            }
+        }
+    }
+
+    /// Reclaims the slots of the SRP host and services that are reported as removed
+    fn cleanup(
+        &mut self,
+        host_info: &otSrpClientHostInfo,
+        mut removed_services: Option<&otSrpClientService>,
+    ) {
+        if host_info.mState == otSrpClientItemState_OT_SRP_CLIENT_ITEM_STATE_REMOVED {
+            self.conf_taken = false;
+            info!("SRP host removed");
+        }
+
+        while let Some(service) = removed_services {
+            let (slot, our_service) = self
+                .services
+                .iter_mut()
+                .enumerate()
+                .find(|(_, s)| core::ptr::eq(&s.service, service))
+                .unwrap();
+
+            removed_services = unsafe { service.mNext.as_ref() };
+
+            our_service.taken = false;
+            info!("SRP service at slot {slot} removed");
+        }
+    }
+
+    fn plat_srp_changed(
+        &mut self,
+        host_info: &otSrpClientHostInfo,
+        _services: Option<&otSrpClientService>,
+        removed_services: Option<&otSrpClientService>,
+    ) {
+        trace!("Plat SRP changed callback");
+
+        self.cleanup(host_info, removed_services);
+
+        // TODO
+        // let state = self.state();
```

Comment:
do you want to clean up this part before merging?

Reply:
Done.